### PR TITLE
Shuffle aswer form

### PR DESCRIPTION
**Pull request by idegtiarov**
add checkbox template and shuffle answers fields in checkbox and multiple_choice templates.

### CODE REVIEW

File: static/js/generic_checkbox.html [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+<head>
+	<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js" type="text/javascript"></script>
+	<script src="http://52.50.241.19/course-content/2017_N/static/js/server_interaction.js" type="text/javascript"></script>
+	<script src="http://d3js.org/d3.v2.min.js?2.10.0"></script>
+	<script>
+
+		function getUrlVars() {
+		  var vars = {};
+		  var url = document.URL;
+		  url = url.split('%24').join('&');
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: Why not `.replace()`?
It looks semantically more correct.

File: static/js/generic_checkbox.html [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+		/*
+			This is a global state dictionary. We define this for convenience when working
+			with edX's state-saving paradigm. Any data that is required for operation to
+			seamlessly persist across page refreshes, or that EdX needs to know about, must
+			be placed in here.
+		*/
+		state = {
+			//is the student's answer correct? (possibly a misnomer... eh)
+				//required for grading the problem
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: indentation

File: static/js/generic_checkbox.html [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+		}
+
+		/*
+			This is a global state dictionary. We define this for convenience when working
+			with edX's state-saving paradigm. Any data that is required for operation to
+			seamlessly persist across page refreshes, or that EdX needs to know about, must
+			be placed in here.
+		*/
+		state = {
+			//is the student's answer correct? (possibly a misnomer... eh)
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: `is_correct` or `is_correct_answer` seems to look more specific.

File: static/js/generic_checkbox.html [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+				//required for grading the problem
+			'correct_answer' : false,
+			//which choice is the answer?
+			'answer' : -1,
+			//list of user inputs
+				//required to repopulate the form so users don't shit bricks
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: Please watch your language.
And indentation.

File: static/js/generic_checkbox.html [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+		//global variable that is set once the form is displayed.
+		//If this is false we can't access the form's document elements,
+		//because they haven't yet been created.
+		displayed = false;
+
+		//url variables
+		vars = {};
+
+		//just stupid rewrapping the url vars into a global variable
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: I know javascript is far from perfect, but there's no need to complain about it in the comments ;)

File: static/js/generic_checkbox.html [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+            	var formstring = '';
+                formstring += '<input type="checkbox" name = "ans" id="ans'+row.toString();
+				formstring += '" onkeypress= "setTimeout(check_answers, 250)"/>';
+				if(div.hasAttribute('choice'+row.toString()))
+				{
+					formstring += div.getAttribute('choice'+row.toString()) + '<br>';
+				}
+				else
+					formstring += '<i>no answer</i><br/>';
+				answerlist.push(formstring);
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: The indentation above looks wrong.

File: static/js/generic_checkbox.html [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+            for (row = 0; row < state['choices']; row++) {
+            	var formstring = '';
+                formstring += '<input type="checkbox" name = "ans" id="ans'+row.toString();
+				formstring += '" onkeypress= "setTimeout(check_answers, 250)"/>';
+				if(div.hasAttribute('choice'+row.toString()))
+				{
+					formstring += div.getAttribute('choice'+row.toString()) + '<br>';
+				}
+				else
+					formstring += '<i>no answer</i><br/>';
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: Add `{}` around `else` block. See link above for samples.

File: static/js/generic_checkbox.html [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        //Compose answer strings for form creation
+        function composeAnswerStringList(div) {
+            var answerlist = [];
+            var row;
+            for (row = 0; row < state['choices']; row++) {
+            	var formstring = '';
+                formstring += '<input type="checkbox" name = "ans" id="ans'+row.toString();
+				formstring += '" onkeypress= "setTimeout(check_answers, 250)"/>';
+				if(div.hasAttribute('choice'+row.toString()))
+				{
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: Please follow code style https://google.github.io/styleguide/jsguide.html#formatting-braces

File: static/js/generic_checkbox.html [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+            var temp = iterObject.slice(0);
+            var cacheString = '';
+            for (steps = iterObject.length; steps > 0; steps--) {
+                var rand = Math.floor(Math.random() * steps);
+                cacheString += temp.splice(rand, 1);
+            }
+            return cacheString;
+        }
+
+		//Creates the form where users input data. Super simple
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: It looks like you're mixing tabs and spaces and the indentation is just explodes. Please recheck that you don't mix them and fix indentation.

File: static/js/generic_checkbox.html (hunk 1 above)
Comment (idegtiarov) [inline static/js/generic_checkbox.html]: Done

File: static/js/generic_checkbox.html (hunk 3 above)
Comment (idegtiarov) [inline static/js/generic_checkbox.html]: Done

Comment (z4y4ts) [inline static/js/generic_checkbox.html]: ok

File: static/js/generic_multiple_choice.html [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+		}
+		
+		//checks to see if input == answers
+		//This runs whenever a user types a number into the form, and when getGrade() runs
+		function check_answers(){
+			if(!displayed)
+				return; //cannot check the form if it doesn't exist
+			
+			var correct = true;
+			var c;
```
Comment (z4y4ts) [inline static/js/generic_multiple_choice.html]: I'm wondering, why do you define loop variable outside the loop?
Here and in other places.

File: static/js/generic_multiple_choice.html [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+			formstring += shuffleString(answerlist);
+			formstring += '</form>';
+			
+			var form = document.getElementById(divname);
+			form.innerHTML = formstring;
+		}
+		
+		//checks to see if input == answers
+		//This runs whenever a user types a number into the form, and when getGrade() runs
+		function check_answers(){
```
Comment (idegtiarov) [inline static/js/generic_multiple_choice.html]: Done

File: static/js/generic_multiple_choice.html [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+			return JSON.stringify(state);
+		}
+		
+		//pass back the whole state to edX
+		function getState(){
+			return JSON.stringify(state);
+		}
+		
+		//fill form in with inputs from state
+		function repopulate_form()
```
Comment (idegtiarov) [inline static/js/generic_multiple_choice.html]: ok

File: static/js/generic_multiple_choice.html [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+
+			}
+		}
+		
+		//This is called right after a page load, with the data that edX has cached for us
+		//We need to use this to correctly reset the page, which includes:
+		//reselecting a user's previous answer and
+		//setting up the state dictionary
+		function setState(){
+			//browser compatibility crap
```
Comment (idegtiarov) [inline static/js/generic_multiple_choice.html]: Done

File: static/js/generic_checkbox.html [hunk 14]
```diff
@@ -0,0 +1,309 @@
+<html>
```
Comment (idegtiarov) [inline static/js/generic_checkbox.html]: Done

File: static/js/generic_checkbox.html [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+			form.innerHTML = formstring;
+		}
+
+		//create list with correct answers
+		function correctAnswerList(){
+			var anslist = [];
+			var row;
+			for (row = 0; row < state['choices']; row++){
+				if (state['answer'].indexOf(row) > -1) {
+					anslist.push(1)
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: I mean `anslist.push(1)`

File: static/js/generic_multiple_choice.html [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+.axis path, .axis line {
+  fill: none;
+  stroke: #000; 
+  shape-rendering: crispEdges;
+}
+
+</style>
+
+<head>
+	<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js" type="text/javascript"></script>
```
Comment (z4y4ts) [inline static/js/generic_multiple_choice.html]: Since we're moving to https it looks like we'll need to use **https** for all external resources as well.

Not in this pull request, though.

File: static/js/generic_multiple_choice.html [hunk 17]
```diff
@@ -0,0 +1,288 @@
+<html>
```
Comment (z4y4ts) [inline static/js/generic_multiple_choice.html]: Indentation. Please use 4 spaces instead of tabs.

File: static/js/generic_checkbox.html [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+				}
+				else {
+					formstring += '<i>no answer</i><br/>';
+				 }
+				answerlist.push(formstring);
+			}
+			return answerlist;
+		}
+
+		//Shuffle answerlist items in composed form for answers
```
Comment (z4y4ts) [inline static/js/generic_checkbox.html]: Where is it? I can't find it.

File: static/js/generic_multiple_choice.html (hunk 17 above)
Comment (idegtiarov) [inline static/js/generic_multiple_choice.html]: Initial file uses tabs and it will mean I change whole file just for improving style that is bad practice.

File: static/js/generic_multiple_choice.html (hunk 16 above)
Comment (idegtiarov) [inline static/js/generic_multiple_choice.html]: Yep it is.

File: static/js/generic_checkbox.html (hunk 15 above)
Comment (idegtiarov) [inline static/js/generic_checkbox.html]: 1 and 0 are numeric representation of True and False correspondingly and used widely.

File: static/js/generic_checkbox.html (hunk 18 above)
Comment (idegtiarov) [inline static/js/generic_checkbox.html]: line 93, suppose you've looked in generic_multiple_choice instead of generic_checkbox will add the same comment there either.
